drm-howto/modeset-atomic.c

```diff
@@ -0,0 +1,1073 @@
+ * object.
+ */
+
+static int set_crtc_property(drmModeAtomicReq *req,
+	struct drm_object *crtc, uint32_t crtc_id,
+	const char *name, uint64_t value)
+{
+	int i, prop_id = -1;
+
+	for (i = 0; i < crtc->props->count_props; i++) {
+		if (!strcmp(crtc->props_info[i]->name, name)) {
+			prop_id = crtc->props_info[i]->prop_id;
+			break;
+		}
+	}
+
+	if (prop_id < 0) {
+                fprintf(stderr, "no crtc property: %s\n", name);
+		return -EINVAL;
+	}
+
+	return drmModeAtomicAddProperty(req, crtc_id, prop_id, value);
+}
+
+/*
+ * set_connector_property() is a new function. It sets a property value to a
+ * connector object.
+ */
+
+static int set_connector_property(drmModeAtomicReq *req,
+	struct drm_object *connector, uint32_t conn_id,
+	const char *name, uint64_t value)
+{
+	int i, prop_id = -1;
+
+	for (i = 0; i < connector->props->count_props; i++) {
+		if (!strcmp(connector->props_info[i]->name, name)) {
+			prop_id = connector->props_info[i]->prop_id;
+			break;
+		}
+	}
+
+	if (prop_id < 0) {
+                fprintf(stderr, "no connector property: %s\n", name);
+		return -EINVAL;
+	}
+
+	return drmModeAtomicAddProperty(req, conn_id, prop_id, value);
+}
+
+/*
+ * set_plane_property is a new function. It sets a property value to a
+ * plane object.
+ */
+
+static int set_plane_property(drmModeAtomicReq *req,
+	struct drm_object *plane, uint32_t plane_id,
+	const char *name, uint64_t value)
+{
+        int i, prop_id = -1;
+
+        for (i = 0 ; i < plane->props->count_props ; i++) {
+                if (!strcmp(plane->props_info[i]->name, name)) {
+                        prop_id = plane->props_info[i]->prop_id;
+                        break;
+                }
+        }
+
+        if (prop_id < 0) {
+                fprintf(stderr, "no plane property: %s\n", name);
+                return -EINVAL;
+        }
+
+        return drmModeAtomicAddProperty(req, plane_id, prop_id, value);
+}
+
+/*
+ * modeset_atomic_modeset() is also new. It is used to atomically set some
+ * properties to our connector, CRTC and plane objects. This function is used to
+ * setup the modeset before we start drawing.
+ *
+ * Notice that we can't perform an atomic commit without a configured
+ * framebuffer. The intention here is not to draw (the framebuffer is
+ * zeroed), we just want to setup everything before we start to draw.
+ *
+ * After we have every property set in our own objects, we call
+ * drmModeAtomicCommit(). This function internally creates a struct
+ * drm_mode_atomic object that receives all the properties we want to change.
+ * In the end, a single IOCTL will be called using this drm_mode_atomic object:
+ * DRM_IOCTL(fd, DRM_IOCTL_MODE_ATOMIC, &atomic)
+ */
+
+static int modeset_atomic_modeset(int fd, struct modeset_dev *dev)
+{
+	drmModeAtomicReq *req;
+	uint32_t blob_id;
+	int ret, flags;
+	struct drm_object *plane = &dev->plane;
+	struct modeset_buf *buf = &dev->bufs[dev->front_buf ^ 1];
+
+	req = drmModeAtomicAlloc();
```

Comment:
Since this function programs just one connector, `drmModeAtomicAlloc()` needs to be moved from here to callers, so that the callers can do multiple connectors in the same commit.

Reply:
Done.